book-content/chapters/10-deriving-types.md

```diff
@@ -0,0 +1,1460 @@
+
+## Indexed Access Types
+
+Indexed access types in TypeScript allow you to access a property of another type. This is similar to how you would access the value of a property in an object at runtime, but instead operates at the type level.
+
+For example, we could use an indexed access type to extract the type of the `title` property from `AlbumDetails`:
+
+```typescript
+interface Album {
+  title: string;
+  artist: string;
+  releaseYear: number;
+}
+```
+
+If we try to use dot notation to access the `title` property from the `Album` type, TypeScript will throw an error:
+
+```ts twoslash
+// @errors: 2713
+interface Album {
+  title: string;
+  artist: string;
+  releaseYear: number;
+}
+
+// ---cut---
+type AlbumTitle = Album.title;
+```
+
+In this case, the error message has a helpful suggestion: use `Album["title"]` to access the type of the `title` property in the `Album` type:
+
+```ts twoslash
+interface Album {
+  title: string;
+  artist: string;
+  releaseYear: number;
+}
+
+// ---cut---
+type AlbumTitle = Album["title"];
+//   ^?
+```
+
+Using this indexed access syntax, the `AlbumTitle` type is equivalent to `string`, because that's the type of the `title` property in the `Album` interface.
+
+This same approach can be used to extract types from a tuple, where the index is used to access the type of a specific element in the tuple:
+
+```typescript
+type AlbumTuple = [string, string, number];
+type AlbumTitle = AlbumTuple[0];
+```
+
+Once again, the `AlbumTitle` will be a `string` type, because that's the type of the first element in the `AlbumTuple`.
+
+### Chaining Multiple Indexed Access Types
+
+Indexed access types can be chained together to access nested properties. This is useful when working with complex types that have nested structures.
+
+For example, we could use indexed access types to extract the type of the `name` property from the `artist` property in the `Album` type:
+
+```typescript
+interface Album {
+  title: string;
+  artist: {
+    name: string;
+  };
+}
+
+type ArtistName = Album["artist"]["name"];
+```
+
+In this case, the `ArtistName` type will be equivalent to `string`, because that's the type of the `name` property in the `artist` object.
+
+### Passing a Union to an Indexed Access Type
+
+If you want to access multiple properties from a type, you might be tempted to create a union type containing multiple indexed accesses:
+
+```typescript
+type Album = {
+  title: string;
+  isSingle: boolean;
+  releaseYear: number;
+};
+
+type AlbumPropertyTypes =
+  | Album["title"]
+  | Album["isSingle"]
+  | Album["releaseYear"];
+```
+
+This will work, but you can do one better - you can pass a union type to an indexed access type directly:
+
+```ts twoslash
+type Album = {
+  title: string;
+  isSingle: boolean;
+  releaseYear: number;
+};
+// ---cut---
+type AlbumPropertyTypes = Album["title" | "isSingle" | "releaseYear"];
+//   ^?
+```
+
+This is a more concise way to achieve the same result.
```

Comment:
you could mention here that the index type gets distributed over the object type in this case